float64 cx
float64 dcx
float64 cy
float64 dcy
uint8 is_left_support
uint8 is_right_support